src/cmap/conn/mod.rs

```diff
@@ -262,13 +290,19 @@ impl Drop for Connection {
         // dropped while it's not checked out. This means that the pool called the `close_and_drop`
         // helper explicitly, so we don't add it back to the pool or emit any events.
         if let Some(ref weak_pool_ref) = self.pool {
-            if let Some(strong_pool_ref) = weak_pool_ref.upgrade() {
-                let dropped_connection_state = self.take();
-                RUNTIME.execute(async move {
-                    strong_pool_ref
-                        .check_in(dropped_connection_state.into())
-                        .await;
-                });
+            // If there's an unfinished write on the connection, then we close the connection rather
+            // than returning it to the pool. We do this because finishing the write could
+            // potentially cause surprising side-effects for the user, who might expect
+            // the operation not to occur due to the future being dropped.
+            if !self.partial_message_state.unfinished_write {
+                if let Some(strong_pool_ref) = weak_pool_ref.upgrade() {
+                    let dropped_connection_state = self.take();
+                    RUNTIME.execute(async move {
+                        strong_pool_ref
+                            .check_in(dropped_connection_state.into())
+                            .await;
+                    });
+                }
             } else {
                 self.close(ConnectionClosedReason::PoolClosed);
```

Comment:
this closed reason needs to be changed, maybe to `Error`, or perhaps to new closed reason case (is this okay with CMAP?). The `PoolClosed` one needs to move to an `else` on the upgrade attempt.